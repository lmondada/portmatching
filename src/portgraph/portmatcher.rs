use std::hash::Hash;

use petgraph::visit::{GraphBase, IntoNodeIdentifiers};

use crate::{
    constraint::ScopeConstraint, pattern, ManyMatcher, Pattern, PatternID, SinglePatternMatcher,
};

use super::{
    pattern::UnweightedPortgraphPattern, EdgeProperty, NodeProperty, PortgraphPattern,
};

/// A pattern matcher on portgraphs.
///
/// Loop through every possible root node in the graph and query a root port
/// matcher for each.
///
/// For graphs that implement petgraph's `IntoNodeIdentifiers` trait, it suffices
/// to implement [`RootedPortMatcher`] and rely on the default implementations
/// of [`PortMatcher`].
pub trait PortMatcher: RootedPortMatcher {
    fn nodes_iter(
        &self,
        graph: GraphRef<Self>,
    ) -> impl Iterator<Item = pattern::Value<Self::Pattern>>;

    /// Find matches of all patterns in `graph`.
    ///
    /// The default implementation loops over all possible `root` nodes and
    /// calls [`PortMatcher::find_anchored_matches`] for each of them.
    fn find_matches(&self, graph: GraphRef<Self>) -> Vec<pattern::Match<Self::Pattern>> {
        let mut matches = Vec::new();
        for root in self.nodes_iter(graph) {
            matches.append(&mut self.find_rooted_matches(graph, root));
        }
        matches
    }
}
/// A pattern matcher for fixed root vertices in port graphs
///
/// A pattern matcher is a type that can find matches of a pattern in a graph.
/// Implement [`RootedPortMatcher::find_rooted_matches`] that finds matches of all
/// patterns rooted at a given root node.
pub trait RootedPortMatcher {
    /// Pattern
    type Pattern: Pattern;

    /// Find matches of all patterns in `graph` anchored at the given `root`.
    fn find_rooted_matches(
        &self,
        graph: pattern::DataRef<Self::Pattern>,
        root: pattern::Value<Self::Pattern>,
    ) -> Vec<pattern::Match<Self::Pattern>>;

    fn get_pattern(&self, id: PatternID) -> Option<&Self::Pattern>;
}

type GraphRef<'g, M> = pattern::DataRef<'g, <M as RootedPortMatcher>::Pattern>;
type NodeId<M> = pattern::Value<<M as RootedPortMatcher>::Pattern>;

impl<U: Eq + Hash + Copy + Ord, M: RootedPortMatcher<Pattern = UnweightedPortgraphPattern<U>>>
    PortMatcher for M
where
    for<'g> GraphRef<'g, Self>: IntoNodeIdentifiers + GraphBase<NodeId = NodeId<Self>>,
{
    fn nodes_iter(
        &self,
        graph: GraphRef<Self>,
    ) -> impl Iterator<Item = pattern::Value<Self::Pattern>> {
        graph.node_identifiers()
    }
}

impl<U, PNode, PEdge> RootedPortMatcher for ManyMatcher<PortgraphPattern<U, PNode, PEdge>>
where
    PortgraphPattern<U, PNode, PEdge>: Pattern,
    <PortgraphPattern<U, PNode, PEdge> as Pattern>::Constraint: ScopeConstraint,
    PEdge: EdgeProperty,
    U: Eq + Hash + Copy,
    PNode: NodeProperty,
{
    type Pattern = PortgraphPattern<U, PNode, PEdge>;

    fn find_rooted_matches(
        &self,
        graph: pattern::DataRef<Self::Pattern>,
        root: pattern::Value<Self::Pattern>,
    ) -> Vec<pattern::Match<PortgraphPattern<U, PNode, PEdge>>> {
        self.run(root, graph)
    }

    fn get_pattern(&self, id: PatternID) -> Option<&Self::Pattern> {
        self.get_pattern(id)
    }
}

impl<U> RootedPortMatcher for ManyMatcher<UnweightedPortgraphPattern<U>>
where
    UnweightedPortgraphPattern<U>: Pattern,
    U: Eq + Hash + Copy,
{
    type Pattern = UnweightedPortgraphPattern<U>;

    fn find_rooted_matches(
        &self,
        graph: pattern::DataRef<Self::Pattern>,
        root: pattern::Value<Self::Pattern>,
    ) -> Vec<pattern::Match<Self::Pattern>> {
        self.run(root, graph)
    }

    fn get_pattern(&self, id: PatternID) -> Option<&Self::Pattern> {
        todo!()
    }
}

impl<P: Pattern> RootedPortMatcher for SinglePatternMatcher<P> {
    type Pattern = P;

    fn find_rooted_matches(
        &self,
        graph: pattern::DataRef<Self::Pattern>,
        root: pattern::Value<Self::Pattern>,
    ) -> Vec<pattern::Match<Self::Pattern>> {
        self.find_rooted_match(root, graph)
    }

    fn get_pattern(&self, id: PatternID) -> Option<&Self::Pattern> {
        assert!(id.0 == 0);
        Some(&self.pattern())
    }
}
