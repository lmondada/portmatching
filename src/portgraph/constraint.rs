//! Constraints for port graph matching.
//!
//! There are two main logic bits
//!  - how to express several constraints as mutually exclusive
//!    constraints; this can be found in the mutex submodule.
//!  - how to decompose a constraint into a list of constraints. Note that this
//!    has to be consistent with how the indexing scheme binds key-values.

mod mutex;

use std::{cmp, collections::BTreeSet, iter};

use itertools::Itertools;
use petgraph::visit::EdgeCount;
use portgraph::{NodeIndex, PortGraph, PortView};

use crate::{
    constraint::DetHeuristic,
    mutex_tree::{ConditionedPredicate, MutuallyExclusiveTree, ToConstraintsTree},
    utils::{portgraph::line_partition, sort_with_indices},
    Constraint, HashMap,
};

use super::{indexing::PGIndexKey, predicate::PGPredicate};
use mutex::*;

/// A constraint on a port graph.
pub type PGConstraint<W = ()> = Constraint<PGIndexKey, PGPredicate<W>>;

impl ToConstraintsTree<PGIndexKey> for PGPredicate {
    fn to_constraints_tree(constraints: Vec<PGConstraint>) -> MutuallyExclusiveTree<PGConstraint> {
        if constraints.is_empty() {
            return MutuallyExclusiveTree::new();
        }
        let constraints = sort_with_indices(constraints);
        // This will always add the first constraint to the tree, plus any other
        // that are mutually exclusive
        mutex_filter(constraints)
    }
}

impl ConditionedPredicate<PGIndexKey> for PGPredicate {
    fn conditioned(
        constraint: &Constraint<PGIndexKey, Self>,
        satisfied: &[&Constraint<PGIndexKey, Self>],
    ) -> Option<Constraint<PGIndexKey, Self>> {
        if !matches!(constraint.predicate(), PGPredicate::IsNotEqual { .. }) {
            return Some(constraint.clone());
        }
        let first_key = constraint.required_bindings()[0];
        let mut keys: BTreeSet<_> = constraint.required_bindings()[1..]
            .iter()
            .copied()
            .collect();
        for s in satisfied
            .iter()
            .filter(|s| s.required_bindings()[0] == first_key)
        {
            for k in s.required_bindings()[1..].iter().copied() {
                keys.remove(&k);
            }
        }
        if keys.is_empty() {
            return None;
        }
        let mut args = vec![first_key];
        let n_other = keys.len();
        args.extend(keys);
        Some(PGConstraint::try_new(PGPredicate::IsNotEqual { n_other }, args).unwrap())
    }
}

impl DetHeuristic for PGConstraint {
    fn make_det(_: &[&Self]) -> bool {
        true
    }
}

impl PartialOrd for PGConstraint {
    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PGConstraint {
    fn cmp(&self, other: &Self) -> cmp::Ordering {
        self.max_key().cmp(&other.max_key())
    }
}

impl PGConstraint {
    /// The key to use for comparisons.
    ///
    /// Constraints are sorted by their largest key binding.
    fn max_key(&self) -> (PGIndexKey, &PGPredicate) {
        let max_key = self
            .required_bindings()
            .iter()
            .max()
            .copied()
            .unwrap_or(PGIndexKey::PathRoot { index: 0 });
        (max_key, self.predicate())
    }
}

pub(super) fn constraint_vec(graph: &PortGraph, root: NodeIndex) -> Vec<PGConstraint> {
    if graph.edge_count() == 0 {
        return vec![PGConstraint::try_new(
            PGPredicate::HasNodeWeight(()),
            vec![PGIndexKey::root(0)],
        )
        .unwrap()];
    }
    let mut constraints = Vec::new();
    let mut node_to_key = HashMap::from_iter([(root, PGIndexKey::root(0))]);
    let mut node_to_root_ind = HashMap::from_iter([(root, 0)]);
    for line in line_partition(&graph, root) {
        let root = graph.port_node(line[0].0).unwrap();
        let n_roots = node_to_root_ind.len();
        let root_index = *node_to_root_ind.entry(root).or_insert(n_roots);
        let root_offset = graph.port_offset(line[0].0).unwrap();
        for (i, (left, right)) in line.into_iter().enumerate() {
            let left_node = graph.port_node(left).unwrap();
            let left_port = graph.port_offset(left).unwrap();
            let right_node = graph.port_node(right).unwrap();
            let right_port = graph.port_offset(right).unwrap();
            let left_key = *node_to_key.get(&left_node).expect("unknown edge LHS");
            let right_key = if let Some(right_key) = node_to_key.get(&right_node) {
                *right_key
            } else {
                // Create a new key
                let key = PGIndexKey::AlongPath {
                    path_root: root_index,
                    path_start_port: root_offset,
                    path_length: i + 1,
                };
                // Ensure it does not clash with previously bound keys
                let args = iter::once(key)
                    .chain(node_to_key.values().copied())
                    .collect_vec();
                constraints.push(
                    PGConstraint::try_new(
                        PGPredicate::IsNotEqual {
                            n_other: node_to_key.len(),
                        },
                        args,
                    )
                    .unwrap(),
                );
                node_to_key.insert(right_node, key);
                key
            };
            constraints.push(
                PGConstraint::try_new(
                    PGPredicate::IsConnected {
                        left_port,
                        right_port,
                    }
                    .into(),
                    vec![left_key, right_key],
                )
                .unwrap(),
            );
        }
    }
    if constraints.is_empty() {
        // We add one (dummy) constraint for the empty pattern, forcing
        // the matcher to bind the first character to a position in the
        // string when matched. An alternative would be to explicitly
        // disallow empty patterns.
        constraints.push(
            PGConstraint::try_new(
                PGPredicate::IsNotEqual { n_other: 0 },
                vec![PGIndexKey::root(0)],
            )
            .unwrap(),
        );
    }
    constraints
}

#[cfg(test)]
mod tests {
    use portgraph::PortOffset;
    use rstest::{fixture, rstest};

    use super::*;

    fn vname(root_id: usize, root_port: Option<PortOffset>, node_index: usize) -> PGIndexKey {
        if let Some(root_port) = root_port {
            PGIndexKey::AlongPath {
                path_root: root_id,
                path_start_port: root_port,
                path_length: node_index,
            }
        } else {
            assert_eq!(root_port, None);
            PGIndexKey::PathRoot { index: root_id }
        }
    }

    fn filter(
        left_port: PortOffset,
        (left_root_id, left_root_port, left_node_index): (usize, Option<PortOffset>, usize),
        right_port: PortOffset,
        (right_root_id, right_root_port, right_node_index): (usize, Option<PortOffset>, usize),
    ) -> PGConstraint {
        PGConstraint::try_new(
            PGPredicate::IsConnected {
                left_port,
                right_port,
            }
            .into(),
            vec![
                vname(left_root_id, left_root_port, left_node_index),
                vname(right_root_id, right_root_port, right_node_index),
            ],
        )
        .unwrap()
    }

    #[fixture]
    fn constraints() -> Vec<PGConstraint> {
        vec![
            filter(
                PortOffset::new_outgoing(0),
                (1, None, 0),
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),
            filter(
                PortOffset::new_incoming(0),
                (2, None, 0),
                PortOffset::new_outgoing(0),
                (0, None, 0),
            ),
            filter(
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_outgoing(0)), 2),
                PortOffset::new_outgoing(0),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),
            filter(
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_outgoing(0)), 2),
                PortOffset::new_outgoing(1),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),
        ]
    }

    #[rstest]
    fn test_to_mut_ex_assign_0(constraints: Vec<PGConstraint>) {
        let tree = PGPredicate::to_constraints_tree(constraints.clone());
        assert_eq!(
            tree.children(tree.root())
                .map(|(_, c)| c.clone())
                .collect_vec(),
            [
                filter(
                    PortOffset::new_incoming(0),
                    (0, Some(PortOffset::new_outgoing(0)), 2),
                    PortOffset::new_outgoing(0),
                    (0, Some(PortOffset::new_incoming(0)), 1)
                ),
                filter(
                    PortOffset::new_incoming(0),
                    (0, Some(PortOffset::new_outgoing(0)), 2),
                    PortOffset::new_outgoing(1),
                    (0, Some(PortOffset::new_incoming(0)), 1),
                ),
            ]
        );
    }

    #[rstest]
    fn test_to_mut_ex_assign_1(constraints: Vec<PGConstraint>) {
        let tree = PGPredicate::to_constraints_tree(constraints[..2].to_vec());
        assert_eq!(
            tree.children(tree.root())
                .map(|(_, c)| c.clone())
                .collect_vec(),
            [filter(
                PortOffset::new_outgoing(0),
                (1, None, 0),
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),]
        );
    }

    #[test]
    fn test_constraint_ordering() {
        let a = filter(
            PortOffset::new_incoming(0),
            (0, Some(PortOffset::new_outgoing(0)), 1),
            PortOffset::new_outgoing(0),
            (0, None, 0),
        );
        let b = filter(
            PortOffset::new_incoming(0),
            (0, Some(PortOffset::new_outgoing(0)), 1),
            PortOffset::new_outgoing(0),
            (1, None, 0),
        );
        // For same literal, an AssignPredicate should be smaller
        assert!(a < b);
    }
}
