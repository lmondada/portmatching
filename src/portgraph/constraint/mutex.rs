//! Logic to compute mutually exclusive PGConstraints.

use crate::{constraint_tree::ConstraintTree, portgraph::predicate::PGPredicate};

use super::PGConstraint;

fn fst_required_binding_eq(a: &PGConstraint, b: &PGConstraint) -> bool {
    a.required_bindings()[0] == b.required_bindings()[0]
}
/// Build a constraint tree for filter constraints.
///
/// Strategy: take the smallest filter (the first), and add all the other filters
/// that are mutually exclusive with it. Filters are mutually exclusive if they
/// are of the same enum variant and
///  - for PGPredicate::IsConnected: they share the same left key and left_port.
///  - for PGPredicate::HasNodeWeight: they share the same key.
///  - for PGPredicate::IsNotEqual: they match on the first key, and if A resp. B
///    are the set of remaining keys (i.e. all but the first argument), Then we
///    create three IsNotEqual constraints, checking for clash with (i) A ∧ B,
///    (ii) A ∧ ¬B and (iii) ¬A ∧ B.
///    We use the fact that constraints are checked in order to simplify this
///    to the equivalent sequence of constraints (i) A ∧ B, (ii) A, (iii) B.
///
/// This assumes that all constraints are distinct and the vec is not empty.
pub(super) fn mutex_filter(
    constraints: Vec<(PGConstraint, usize)>,
) -> ConstraintTree<PGConstraint> {
    // Assume constraints is not empty
    let first_constraint = constraints[0].0.clone();

    match first_constraint.predicate() {
        PGPredicate::IsNotEqual { .. } => {
            let constraints = constraints.into_iter().filter(|(c, _)| {
                // We can only turn IsNotEqual constraints into mutex predicates
                // if they act on the same variable
                matches!(c.predicate(), PGPredicate::IsNotEqual { .. })
                    && fst_required_binding_eq(c, &first_constraint)
            });
            ConstraintTree::with_powerset(constraints.collect())
        }
        PGPredicate::HasNodeWeight(..) | PGPredicate::IsConnected { .. } => {
            ConstraintTree::with_transitive_mutex(constraints, |a, b| {
                match (a.predicate(), b.predicate()) {
                    (PGPredicate::HasNodeWeight(..), PGPredicate::HasNodeWeight(..)) => {
                        fst_required_binding_eq(a, b)
                    }
                    (
                        PGPredicate::IsConnected {
                            left_port: lp_a, ..
                        },
                        PGPredicate::IsConnected {
                            left_port: lp_b, ..
                        },
                    ) => lp_a == lp_b && fst_required_binding_eq(a, b),
                    _ => false,
                }
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use portgraph::PortOffset;

    use crate::portgraph::indexing::PGIndexKey;

    use super::*;

    fn index_key(path_root: usize, path_length: usize) -> PGIndexKey {
        PGIndexKey::AlongPath {
            path_root,
            path_start_port: PortOffset::Outgoing(0),
            path_length,
        }
    }

    #[test]
    fn test_all_combinations() {
        let constraints = vec![
            (
                PGConstraint::try_new(
                    PGPredicate::<()>::IsNotEqual { n_other: 2 },
                    vec![index_key(0, 0), index_key(1, 1), index_key(2, 1)],
                )
                .unwrap(),
                0,
            ),
            (
                PGConstraint::try_new(
                    PGPredicate::IsNotEqual { n_other: 2 },
                    vec![index_key(0, 0), index_key(1, 1), index_key(1, 2)],
                )
                .unwrap(),
                1,
            ),
            (
                PGConstraint::try_new(
                    PGPredicate::IsNotEqual { n_other: 2 },
                    vec![index_key(0, 0), index_key(1, 1), index_key(1, 3)],
                )
                .unwrap(),
                2,
            ),
        ];
        let tree = ConstraintTree::with_powerset(constraints);
        assert_eq!(tree.n_nodes(), 1 + 3 + 2 + 1 + 1);
        assert_debug_snapshot!(tree);
    }
}
