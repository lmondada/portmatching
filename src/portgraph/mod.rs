mod constraint;
pub(crate) mod pattern;
mod portmatcher;
mod weighted_graph;

use constraint::PortgraphConstraint;
use portgraph::{PortGraph, Weights};

pub use self::portmatcher::{PortMatcher, RootedPortMatcher};
pub use pattern::{PortgraphPattern, PortgraphPatternBuilder};
pub use weighted_graph::WeightedPortGraphRef;

use std::hash::Hash;
use std::iter::{self, Map, Repeat, Zip};
use std::ops::RangeFrom;

pub trait EdgeProperty: Clone + Ord + Hash {
    type OffsetID: Eq + Copy;

    fn reverse(&self) -> Option<Self>;

    fn offset_id(&self) -> Self::OffsetID;
}


pub trait NodeProperty: Clone + Hash + Ord {}

impl<U: Clone + Hash + Ord> NodeProperty for U {}

impl<A: Copy + Ord + Hash> EdgeProperty for (A, A) {
    type OffsetID = A;

    fn reverse(&self) -> Option<Self> {
        (self.1, self.0).into()
    }

    fn offset_id(&self) -> Self::OffsetID {
        self.0
    }
}

impl EdgeProperty for () {
    type OffsetID = ();

    fn reverse(&self) -> Option<Self> {
        ().into()
    }

    fn offset_id(&self) -> Self::OffsetID {}
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Symbol(IterationStatus, usize);
type SymbolsIter =
    Map<Zip<Repeat<IterationStatus>, RangeFrom<usize>>, fn((IterationStatus, usize)) -> Symbol>;

impl Symbol {
    pub(crate) fn new(status: IterationStatus, ind: usize) -> Self {
        Self(status, ind)
    }

    fn from_tuple((status, ind): (IterationStatus, usize)) -> Self {
        Self::new(status, ind)
    }

    pub(crate) fn root() -> Self {
        Self(IterationStatus::Skeleton(0), 0)
    }

    pub(crate) fn symbols_in_status(status: IterationStatus) -> SymbolsIter {
        iter::repeat(status).zip(0..).map(Self::from_tuple)
    }
}

impl crate::Symbol for Symbol {
    fn root() -> Self {
        Symbol::root()
    }
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
enum IterationStatus {
    // We are traversing the i-th line wihtin skeleton
    Skeleton(usize),
    // We are traversing the i-th line outside skeleton
    LeftOver(usize),
    // We are done
    Finished,
}

impl IterationStatus {
    fn increment(&mut self, max_i: usize) {
        *self = match self {
            Self::Skeleton(i) => {
                if *i + 1 < max_i {
                    Self::Skeleton(*i + 1)
                } else {
                    Self::LeftOver(0)
                }
            }
            Self::LeftOver(i) => {
                if *i + 1 < max_i {
                    Self::LeftOver(*i + 1)
                } else {
                    Self::Finished
                }
            }
            Self::Finished => Self::Finished,
        }
    }
}


#[derive(PartialEq, Eq, Copy, Clone, Hash, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
enum NodeLocation {
    // The node is in an already-known location
    Exists(Symbol),
    // We need to explore along the i-th line to discover the node
    Discover(usize),
}