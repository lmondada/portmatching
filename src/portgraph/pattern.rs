//! Patterns for matching port graphs.

use crate::{pattern::ConcretePattern, utils::is_connected, Pattern};
use portgraph::{LinkView, NodeIndex, PortGraph};

use super::constraint::{constraint_vec, PGConstraint, PGPatternError};

/// A concrete port graph pattern.
///
/// This has no variables, and thus can be represented by a port graph directly.
#[derive(Clone, PartialEq, Eq, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct PGPattern<G> {
    pub(super) graph: G,
    pub(super) root: Option<NodeIndex>,
}

impl Pattern for PGPattern<PortGraph> {
    type Constraint = PGConstraint;
    type Error = PGPatternError;

    fn try_to_constraint_vec(&self) -> Result<Vec<Self::Constraint>, Self::Error> {
        constraint_vec(&self.graph, self.root.ok_or(PGPatternError::NoRoot)?)
    }
}

impl ConcretePattern for PGPattern<PortGraph> {
    type Host = PortGraph;

    fn as_host(&self) -> &Self::Host {
        &self.graph
    }
}

/// Converting a pattern to constraints requires setting a root first.
#[derive(Debug)]
pub struct NoRootFound;

impl<G> PGPattern<G> {
    /// Whether a root has been set.
    pub fn is_root_set(&self) -> bool {
        self.root.is_some()
    }

    /// Set the pattern root.
    pub fn set_root(&mut self, root: NodeIndex) {
        self.root = Some(root);
    }

    /// Construct a pattern from a port graph, without setting root.
    pub fn from_host(graph: G) -> Self {
        Self { graph, root: None }
    }

    /// Construct a pattern from a port graph, use the default root choice.
    pub fn from_host_pick_root(graph: G) -> Self
    where
        G: LinkView,
    {
        let mut p = Self { graph, root: None };
        p.pick_root().unwrap();
        p
    }

    /// Construct a pattern given a port graph and node as root.
    pub fn from_host_with_root(graph: G, root: NodeIndex) -> Self {
        Self {
            graph,
            root: Some(root),
        }
    }
}

impl<G: LinkView> PGPattern<G> {
    /// Whether the pattern is connected.
    ///
    /// A pattern must always be connected for pattern matching.
    pub fn is_connected(&self) -> bool {
        is_connected(&self.graph)
    }

    /// Attempt to pick a valid root.
    pub fn pick_root(&mut self) -> Result<NodeIndex, NoRootFound> {
        let root = self.graph.nodes_iter().next().ok_or(NoRootFound)?;
        self.root = Some(root);
        Ok(root)
    }
}
