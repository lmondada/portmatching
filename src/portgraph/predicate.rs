//! Predicates (the core of constraints), for port graph matching.

use std::{borrow::Borrow, fmt::Debug};

use itertools::Itertools;
use portgraph::{LinkView, NodeIndex, PortGraph, PortOffset, SecondaryMap, UnmanagedDenseMap};

use crate::predicate::{ArityPredicate, Predicate};

/// A predicate for constraints on (weighted) port graph
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum PGPredicate<NodeWeight = ()> {
    /// The node has a specific weight
    HasNodeWeight(NodeWeight),
    /// The two nodes are connected by an edge from `left_port` to `right_port`
    IsConnected {
        /// The left port of the edge
        left_port: PortOffset,
        /// The right port of the edge
        right_port: PortOffset,
    },
    /// The node is not equal to any of the other `n_other` nodes
    IsNotEqual {
        /// The number of other nodes, determining the predicate arity.
        n_other: usize,
    },
}

impl<W: Eq + Debug> PGPredicate<W> {
    /// Whether the predicate is satisfied, given graph, node weights and arguments
    fn is_satisfied(
        &self,
        graph: &PortGraph,
        node_weights: &impl SecondaryMap<NodeIndex, W>,
        args: &[impl Borrow<NodeIndex>],
    ) -> bool {
        match self {
            PGPredicate::HasNodeWeight(weight) => {
                let [node] = args else {
                    panic!("expected args of length 1 (arity 1)")
                };
                node_weights.get(*node.borrow()) == weight
            }
            &PGPredicate::IsConnected {
                left_port,
                right_port,
            } => {
                let [left, right] = args else {
                    panic!("expected args of length 2 (arity 2)")
                };
                has_edge(
                    graph,
                    *left.borrow(),
                    left_port,
                    *right.borrow(),
                    right_port,
                )
            }
            &PGPredicate::IsNotEqual { n_other } => {
                let [node, other @ ..] = args else {
                    panic!("expected args of length {}", n_other + 1)
                };
                let other = other.iter().map(|n| *n.borrow()).collect_vec();
                !other.contains(node.borrow())
            }
        }
    }
}

impl<W: Clone + Eq> ArityPredicate for PGPredicate<W> {
    fn arity(&self) -> usize {
        match self {
            PGPredicate::HasNodeWeight(..) => 1,
            PGPredicate::IsConnected { .. } => 2,
            PGPredicate::IsNotEqual { n_other } => n_other + 1,
        }
    }
}

impl Predicate<PortGraph> for PGPredicate {
    type Value = NodeIndex;

    fn check(&self, data: &PortGraph, args: &[impl Borrow<Self::Value>]) -> bool {
        self.is_satisfied(data, &UnmanagedDenseMap::default(), args)
    }
}

impl<M, W> Predicate<(&PortGraph, &M)> for PGPredicate<W>
where
    M: SecondaryMap<NodeIndex, W>,
    W: Clone + Eq + Debug,
{
    type Value = NodeIndex;

    fn check(&self, data: &(&PortGraph, &M), args: &[impl Borrow<Self::Value>]) -> bool {
        let &(graph, weights) = data;
        self.is_satisfied(graph, weights, args)
    }
}

impl<W: Debug> Debug for PGPredicate<W> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PGPredicate::HasNodeWeight(weight) => write!(f, "has_node_weight({:?})", weight),
            PGPredicate::IsConnected {
                left_port,
                right_port,
                ..
            } => write!(f, "{:?} -> {:?}", left_port, right_port),
            PGPredicate::IsNotEqual { n_other } => write!(f, "not_equal({})", n_other),
        }
    }
}

fn has_edge<G: LinkView>(
    graph: &G,
    left: NodeIndex,
    left_port: PortOffset,
    right: NodeIndex,
    right_port: PortOffset,
) -> bool {
    let Some(left_port) = graph.port_index(left, left_port) else {
        return false;
    };
    graph
        .port_links(left_port)
        .filter(|&(_, p)| {
            graph.port_offset(p) == Some(right_port) && graph.port_node(p) == Some(right)
        })
        .next()
        .is_some()
}
