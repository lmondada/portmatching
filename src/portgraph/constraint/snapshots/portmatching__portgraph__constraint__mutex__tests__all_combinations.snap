---
source: src/portgraph/constraint/mutex.rs
expression: tree
---
MutuallyExclusiveTree {
    nodes: [
        MutExTreeNode {
            constraint_indices: [],
            children: [
                MutExTreeNodeChild {
                    predicate: not_equal(2)(Along(0, out0)@0), Along(1, out0)@1), Along(2, out0)@1)),
                    node_index: 1,
                },
                MutExTreeNodeChild {
                    predicate: not_equal(2)(Along(0, out0)@0), Along(1, out0)@1), Along(1, out0)@2)),
                    node_index: 2,
                },
                MutExTreeNodeChild {
                    predicate: not_equal(2)(Along(0, out0)@0), Along(1, out0)@1), Along(1, out0)@3)),
                    node_index: 3,
                },
            ],
        },
        MutExTreeNode {
            constraint_indices: [
                0,
            ],
            children: [
                MutExTreeNodeChild {
                    predicate: not_equal(1)(Along(0, out0)@0), Along(1, out0)@2)),
                    node_index: 4,
                },
                MutExTreeNodeChild {
                    predicate: not_equal(1)(Along(0, out0)@0), Along(1, out0)@3)),
                    node_index: 5,
                },
            ],
        },
        MutExTreeNode {
            constraint_indices: [
                1,
            ],
            children: [
                MutExTreeNodeChild {
                    predicate: not_equal(1)(Along(0, out0)@0), Along(1, out0)@3)),
                    node_index: 6,
                },
            ],
        },
        MutExTreeNode {
            constraint_indices: [
                2,
            ],
            children: [],
        },
        MutExTreeNode {
            constraint_indices: [
                1,
            ],
            children: [
                MutExTreeNodeChild {
                    predicate: not_equal(2)(Along(0, out0)@0), Along(1, out0)@1), Along(1, out0)@3)),
                    node_index: 7,
                },
            ],
        },
        MutExTreeNode {
            constraint_indices: [
                2,
            ],
            children: [],
        },
        MutExTreeNode {
            constraint_indices: [
                2,
            ],
            children: [],
        },
        MutExTreeNode {
            constraint_indices: [
                2,
            ],
            children: [],
        },
    ],
}
